File: possible-bipartition/UnionFind.go
package possible_bipartition

type UnionFind struct {
	parents, sizes map[int]int
}

func (u UnionFind) Size(x int) int {

	v, o := u.sizes[x]
	if o {
		return v
	} else {
		return 1
	}
}
func NewUnionFind() UnionFind {
	parent := map[int]int{}

	return UnionFind{parents: parent, sizes: map[int]int{}}
}

func (u UnionFind) Find(x int) int {
	_, o := u.parents[x]
	if !o {
		u.parents[x] = x
	}
	if u.parents[x] != x {
		u.parents[x] = u.Find(u.parents[x])
	}
	return u.parents[x]
}

func (u UnionFind) Union(x, y int) {
	x, y = u.Find(x), u.Find(y)
	if x == y {
		return
	}

	u.sizes[x] = u.Size(x)
	u.sizes[y] = u.Size(y)

	if u.sizes[x] > u.sizes[y] {
		u.parents[y] = x
		u.sizes[x] = u.sizes[x] + u.sizes[y]
	} else {
		u.parents[x] = y
		u.sizes[y] = u.sizes[x] + u.sizes[y]
	}
}

func (u UnionFind) Connected(x, y int) bool {
	return u.Find(x) == u.Find(y)
}

File: possible-bipartition/UnionFind.kt
package com.github.masx200.leetcode_test.possible_bipartition

class UnionFind<T> {
    fun union(a: T, b: T) {
        val fa = this.find(a)
        val fb = this.find(b)
        if (fa == fb) {
            return
        }
        val sa = this.sizes[fa] ?: 1
        val sb = this.sizes[fb] ?: 1

        if (sa < sb) {
            this.parents[fa] = fb
            this.sizes[fb] = sb + sa
        } else {
            this.parents[fb] = fa
            this.sizes[fa] = sb + sa
        }
    }

    private val sizes = mutableMapOf<T, Int>()
    private val parents = mutableMapOf<T, T>()
    fun connected(p: T, q: T): Boolean {
        return this.find(p) == this.find(q)
    }

    fun find(x: T): T {
        if (x != (this.parents[x] ?: x)) {
            this.parents[x] = this.find(this.parents[x] ?: x)

        }
        return this.parents[x] ?: x
    }

    fun size(x: T): Int {
        return this.sizes.getOrDefault(x, 1)

    }
}